amaranth/lib/data.py

```diff
@@ -0,0 +1,395 @@
+from abc import ABCMeta, abstractmethod, abstractproperty
+from collections.abc import Mapping
+
+from amaranth.hdl import *
+from amaranth.hdl.ast import ShapeCastable, ValueCastable
+
+
+__all__ = [
+    "Field", "Layout", "StructLayout", "UnionLayout", "ArrayLayout", "IrregularLayout",
+    "View", "Struct", "Union",
+]
+
+
+class Field:
+    def __init__(self, shape, offset):
+        self.shape  = shape
+        self.offset = offset
+
+    @property
+    def shape(self):
+        return self._shape
+
+    @shape.setter
+    def shape(self, shape):
+        try:
+            Shape.cast(shape)
+        except TypeError as e:
+            raise TypeError("Field shape must be a shape-castable object, not {!r}"
+                            .format(shape)) from e
+        self._shape = shape
+
+    @property
+    def offset(self):
+        return self._offset
+
+    @offset.setter
+    def offset(self, offset):
+        if not isinstance(offset, int) or offset < 0:
+            raise TypeError("Field offset must be a non-negative integer, not {!r}"
+                            .format(offset))
+        self._offset = offset
+
+    @property
+    def width(self):
+        return Shape.cast(self.shape).width
+
+    def __eq__(self, other):
+        return (isinstance(other, Field) and
+                self._shape == other.shape and self._offset == other.offset)
+
+    def __repr__(self):
+        return f"Field({self._shape!r}, {self._offset})"
+
+
+class Layout(ShapeCastable, metaclass=ABCMeta):
+    @staticmethod
+    def cast(obj):
+        """Cast a shape-castable object to a layout."""
+        while isinstance(obj, ShapeCastable):
+            if isinstance(obj, Layout):
+                return obj
+            new_obj = obj.as_shape()
+            if new_obj is obj:
+                break
+            obj = new_obj
+        Shape.cast(obj) # delegate non-layout-specific error handling to Shape
+        raise TypeError("Object {!r} cannot be converted to a data layout"
+                        .format(obj))
+
+    @staticmethod
+    def of(obj):
+        """Extract the layout from a view."""
+        if not isinstance(obj, View):
+            raise TypeError("Object {!r} is not a data view"
+                            .format(obj))
+        return obj._View__orig_layout
```

Comment:
Do you have a name for it that you like?

Reply:
It seems like it could could literally be called `Layout.view`

Reply:
There are intentionally as few members with non-reserved names defined on `Layout` because `Layout` forwards accesses to unknown members to the fields defined in structures. If I add a member called `view`, then defining a structure with a `view` field would not really be possible.

Right now the same is already true for `eq`, but that's the sole exception and I think it's justified; `view` is harder to justify.

Reply:
Constructor on View that takes a layout?  Possibly even looking for a __view__ member?

Reply:
The `View` constructor already takes a `layout` argument, did you mean something else?

Also, dunder names like `__view__` are reserved by Python itself.

Reply:
I suspect I misunderstood the discussion then, sorry :-)

Reply:
Actually no, it was me who misunderstood the discussion and swapped "layout" and "view" here. Updated response below.

<hr>

We could add `Layout.view`, though since `View` already has a constructor that takes a `Layout` this would be an optional addition. I'm not sure I see many use cases for it; I think that if people like the idea then a new mini-RFC could be written that would argue for it and it doesn't have to be a part of this one.